Teach_Breakout.py

```diff
@@ -0,0 +1,344 @@
+    next_Q_values = target_model.predict([next_states, np.ones((number_in_batch, 4))])
+    # The Q values of the terminal states is 0 by definition:
+    next_Q_values[is_terminal] = 0
+    # The Q values of each start state is the reward + gamma * the max next state Q value:
+    Q_values = rewards + gamma * np.max(next_Q_values, axis=1)
+    # TODO: Doublecheck: correct axis for max? And that next_Q_values gives 4 values (one for each action) for each item in the batch?
+
+    # Turn the actions into one-hot-encoded actions (required to use the mask)
+    one_hot_actions = np.zeros((number_in_batch, 4))
+    for i in range(number_in_batch):
+        for j in range(4):
+            if j == actions[i]:
+                one_hot_actions[i, j] = 1
+
+    # Fit the keras model, using the mask
+    model.fit([start_states, one_hot_actions], one_hot_actions * Q_values[:, None],
+        epochs=1, batch_size=len(start_states), verbose=0) # Should it be epochs=1??
+
+
+def make_n_fit_batch(model, target_model, gamma, iteration, mem_size, mem_frames, mem_actions, mem_rewards, mem_is_terminal, number_in_batch):
+    """Make a batch then use it to train the model"""
+
+    if iteration < mem_size:
+        # In this case, sample 'number_in_batch' numbers from 4 to 'iterations'
+        # (We use 4 so that there are 5 frames minimum, so we can still make 2 states (because each state is the frame
+        # plus the previous 3 frames))
+        indices_chosen = random.sample(range(4, iteration), number_in_batch)
+    else:
+        # Now the memory is full, so we can take any elements from it
+        indices_chosen = random.sample(range(0, mem_size), number_in_batch)
+
+    # Initialise the batches
+    start_states = np.zeros((number_in_batch, 105, 80, 4))  # (Array instead of list)
+    next_states = np.zeros((number_in_batch, 105, 80, 4))  # (Array instead of list)
+    actions = np.zeros((number_in_batch))  # (Array instead of list)
+
+    rewards = list()  # List
+    is_terminals = list()  # List
+
+    for i in range(len(indices_chosen)): # Probably more efficient way to do this: "for index in ?(indices_chosen)?"
+        index = indices_chosen[i] # index corresponds to the iterations
+        start_states[i, :, :, :] = make_state(mem_frames, index - 1)
+        next_states[i, :, :, :] = make_state(mem_frames, index) # State given by index was arrived at by taking action given by index and reward received is give by index.
+                                                                # In contrast, index-1 labels the previous state, before the action was taken
+        actions[i] = mem_actions.recall(index)
+
+        rewards.append(mem_rewards.recall(index))
+        is_terminals.append(mem_is_terminal.recall(index))
+
+    # We should now have a full batch, which the DNN can train on
+    fit_batch_target(model, target_model, gamma, start_states, actions, rewards, next_states, is_terminals, number_in_batch)
```

Comment:
Avoid lines this long. Convention is should line-break after 80 characters, although I'm in favour of longer line widths (I use 100, some projects use 120). Important is to pick a width and stick to it.